misk/src/main/kotlin/misk/moshi/wire/MapAdapter.kt

```diff
@@ -0,0 +1,43 @@
+package misk.moshi.wire
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+
+/** Handles JSON marshalling / unmarshalling for type-safe maps */
+internal class MapAdapter(
+        private val keyConverter: (String) -> Any,
+        private val valueAdapter: JsonAdapter<Any>
+) : JsonAdapter<Map<Any, Any>>() {
+    override fun fromJson(reader: JsonReader): Map<Any, Any>? {
+        val nextToken = reader.peek()
+        if (nextToken == JsonReader.Token.NULL) return reader.nextNull()
+        check(nextToken == JsonReader.Token.BEGIN_OBJECT) {
+            "expected {, found $nextToken"
+        }
+
+        val results = mutableMapOf<Any, Any>()
+        reader.beginObject()
+        while (reader.peek() != JsonReader.Token.END_OBJECT) {
+            val keyName = reader.nextName()
+            val key = keyConverter(keyName)
```

Comment:
Off-topic: this syntax makes perfect sense but it's still surprising. I'm still learning apparently!